### PR TITLE
[NO SQUASH] Add callback on_mapblocks_changed

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
This could close https://github.com/minetest/minetest/issues/12725. It provides a new callback registered with `register_on_mapblocks_changed`. Basically, all the map modifications each server step are collected and passed to Lua as a set. No modifications are missed. You can read more in lua_api.txt.

Unlike other possible solutions to https://github.com/minetest/minetest/issues/12725, this one might raise many false positives, but it is relatively easy to implement and fast.

## To do

This PR is Ready for Review.

## How to test

I added a Lua unit test.

Here is a mod which uses the functionality. It is a more limited version of the Mesecons node detector. The code is crude and unoptimized.
[mesecons_detector2.zip](https://github.com/minetest/minetest/files/9483706/mesecons_detector2.zip)

### CODE REVIEW

**Comment by Desour** [comment]
#12725 wants to record all map changes, probably even in fine granularity. (I assume the "reply" thing is mentioned is the same as in this forum post: <https://forum.minetest.net/viewtopic.php?f=47&t=28594>.) So I kinda doubt this PR's feature is so helpful for that use-case.

That being said, this feature can still be useful. But:
* For this being useful, it needs to be documented what sort of mapblock changes are detected. I.e. does it include objects that leave or enter the mapblock? This is also important because of backwards compatibility.
* Maybe make the feature mapblock-agnostic. I.e. allow to watch arbitrary cuboid areas: `local handle = minetest.watch_area(pos1, pos2, func)`, `minetest.unwatch_area(handle)`

**Comment by TurkeyMcMac** [comment]
Events are tracked that change nodes or node metadata. I will document this sometime.

I think tracking mapblocks is practical for a callback of this nature, since it doesn't get too expensive even with bulk changes like those done by Mesecons. For the `minetest.watch_area` API you propose, we could use an `AreaStore` type thing, with each area corresponding to a `minetest.watch_area` call. Map edit events already have a `getArea` method.

File: src/script/common/c_converter.h
Comment (nerzhul) [inline src/script/common/c_converter.h:128]: just a question, cannot we expose this directly instead of duplicating implementation on both side ? what is the cost ?

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/script/common/c_converter.h
Comment (appgurueu) [reply src/script/common/c_converter.h:128]: The cost is that LuaJIT is unable to optimize this if the C++ func is used.

Comment (TurkeyMcMac) [reply src/script/common/c_converter.h:128]: In a later PR someone could look into calling the C++ implementation using LuaJIT's FFI library.

File: doc/lua_api.txt
Comment (sfan5) [inline doc/lua_api.txt:5327]: would be useful to mention `get_position_from_hash` here

File: doc/lua_api.txt
Comment (sfan5) [inline doc/lua_api.txt:5325]: - the mapblock positions are obviously of the ones that were modified, but this is never spelled out
- no modder knows what a "map edit event" is, not sure why this is being mentioned
- possibility of false-positives not mentioned
  (may trigger despite no observable changes from Lua)

File: games/devtest/mods/unittests/misc.lua
Comment (sfan5) [inline games/devtest/mods/unittests/misc.lua:120]: technically this is a no-op that happens to trigger a block change anyway
I think you rather want to test `add_node` or such

File: src/script/common/c_converter.h
Comment (sfan5) [inline src/script/common/c_converter.h:126]:
```suggestion
// Note that this returns a floating point result as Lua integers are 32-bit
inline lua_Number hash_node_position(v3s16 pos)
```

File: src/serverenvironment.cpp
Comment (sfan5) [inline src/serverenvironment.cpp:488]: question is: is it okay if registering a new callback at runtime silently fails?
IIRC no other callbacks behave this way.

File: doc/lua_api.txt
Comment (TurkeyMcMac) [reply doc/lua_api.txt:5329]: The only way I know of would be to loop through all the entries and count up, which would be slower than this way.

File: src/serverenvironment.cpp
Comment (TurkeyMcMac) [reply src/serverenvironment.cpp:488]: It is different, but I don't want people to pay the price of accumulating changed mapblocks if no mod uses the information. If a modder encounters this quirk, they can work around it. I could document it.

Comment (TurkeyMcMac) [reply src/serverenvironment.cpp:488]: I have now added this to the documentation.

Comment (appgurueu) [reply src/serverenvironment.cpp:488]: Would it be too expensive to make this an error? "register_on_mapblocks changed must be called at load time"

**Comment by SmallJoker** [comment]
https://irc.minetest.net/minetest-dev/2022-10-09#i_6023230

Use-cases seem rather limited and yet it would be the only way to reliably catch mapblock writes.

**Comment by TurkeyMcMac** [comment]
I should mention this does not track changes in node timers, since changing them does not create a `MapEditEvent` (or even mark the block as modified, AFAICT.)

**Comment by BuckarooBanzay** [comment]
> https://irc.minetest.net/minetest-dev/2022-10-09#i_6023230
> 
> Use-cases seem rather limited and yet it would be the only way to reliably catch mapblock writes.

I actually have a use-case: in my `modgen` mod i have an "autosave" feature which exports all changed blocks to the disk, 
currently i'm using a couple overrides and other stuff i don't really like: https://github.com/BuckarooBanzay/modgen/blob/master/autosave.lua#L58-L177
With this callback i could replace all of them in one go

File: src/serverenvironment.cpp
Comment (TurkeyMcMac) [reply src/serverenvironment.cpp:488]: I have now done as you suggest (a little while ago actually.)

File: src/serverenvironment.cpp
Comment (nerzhul) [inline src/serverenvironment.cpp:1602]: why we do a swap here and not just pass a const ref ?

File: src/map.h
Comment (nerzhul) [inline src/map.h:92]: this is not required anymore ?

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/map.h
Comment (TurkeyMcMac) [reply src/map.h:92]: It seems to me that a metadata change only affects one node, not an entire block.

File: src/serverenvironment.cpp
Comment (TurkeyMcMac) [reply src/serverenvironment.cpp:1602]: The internal set of modified blocks has to be cleared before calling the Lua callbacks so it is ready to record any changes made by them. The set could be cleared by the `on_mapblocks_changed` function, but this API is cleaner.

**Comment by Zughy** [comment]
@TurkeyMcMac rebase needed

**Comment by sfan5** [comment]
my personal testing code:
```lua
minetest.register_on_mapblocks_changed(function(blocks, block_count)
	if block_count == 0 then
		return
	end
	print(block_count .. " blocks have been modified")
	for _, player in ipairs(core.get_connected_players()) do
		local pos = vector.floor(vector.divide(player:get_pos(), 16))
		if blocks[minetest.hash_node_position(pos)] then
			print("  including the block " .. core.pos_to_string(pos) ..
				" " .. player:get_player_name() .. " was standing in")
		end
	end
end)
```

File: src/server/serverinventorymgr.cpp
Comment (TurkeyMcMac) [reply src/server/serverinventorymgr.cpp]: I added a helper method `MapEditEvent::setPositionModified`.

File: src/server.cpp
Comment (TurkeyMcMac) [reply src/server.cpp:1263]: Yes, as you can see here: https://github.com/minetest/minetest/blob/504e43e0dae50ad98e19db0649c9c825cf2ab7a7/src/server.cpp#L944-L949

**Review by sfan5 — APPROVED** [review]

**Comment by sfan5** [comment]
@TurkeyMcMac this looks ready, will you merge it?

**Comment by TurkeyMcMac** [comment]
I was waiting for another review, but I think it's fine to merge.

**Pull request merged**
